File: [0160][Easy][Intersection_of_Two_Linked_Lists][LinkedList]/Intersection_of_Two_Linked_Lists.java
class ListNode {
    int val;
    ListNode next;
    ListNode(int x) {
        val = x;
        next = null;
    }
}

// leetcode time     cost : 1 ms
// leetcode memory   cost : 42.4 MB
// Time  Complexity: O(m+n)
// Space Complexity: O(1)
 // solution 3, 2 pointers
public class Solution {
    public ListNode getIntersectionNode(ListNode headA, ListNode headB) {
        if (headA == null || headB == null) return null;
        ListNode pA = headA, pB = headB;
        while (pA != pB) {
            pA = pA == null ? headB : pA.next;
            pB = pB == null ? headA : pB.next;
        }
        return pA;
    }
}

File: [0160][Easy][Intersection_of_Two_Linked_Lists][LinkedList]/Intersection_of_Two_Linked_Lists.c

struct ListNode {
    int val;
    struct ListNode next;
};


// leetcode time     cost : 68 ms
// leetcode memory   cost : 13.5 MB
// Time  Complexity: O(m+n)
// Space Complexity: O(1)
 // solution 3, 2 pointers
struct ListNode getIntersectionNode(struct ListNode headA, struct ListNode headB) {
    if(!headA || !headB) return NULL;
    struct ListNode p, q, pp, qq;
    p = headA; q = headB;
    while(p != q)
    {
        pp = p; qq = q;
        p = (!p -> next && qq -> next) ? headB : p -> next;
        q = (!q -> next && pp -> next) ? headA : q -> next;
    }
    return p;
}